### PR TITLE
Initial buddy allocator for gni provider

### PR DESCRIPTION
**Pull request by e-harvey**
Tested and verified `_gnix_buddy_allocator_create`, `_gnix_buddy_allocator_destroy`, `_gnix_buddy_alloc`, and `_gnix_buddy_free`.

Added a fastlock to the allocator handle but did not test or verify the buddy allocator in with a multi-threaded application.

@hppritcha @jswaro @sungeunchoi @ztiffany

### CODE REVIEW

File: prov/gni/include/gnix_buddy_allocator.h [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	size_t len;
+	size_t max;
+
+	size_t nlists;
+	struct dlist_entry *lists;
+
+	gnix_bitmap_t bitmap;
+
+	fastlock_t lock;
+} handle_t;
```
Comment (ztiffany) [inline prov/gni/include/gnix_buddy_allocator.h]: I think handle_t is too generic a type name.

File: prov/gni/src/gnix_buddy_allocator.c [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	/* Ensure parameters are valid */
+	if (!base || !len || !max || max > len || !alloc_handle ||
+	    IS_NOT_POW_TWO(max) || (len % max) || !(len / MIN_BLOCK_SIZE * 2)) {
+
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid parameter to _gnix_buddy_allocator_create."
+			  "\n");
```
Comment (ztiffany) [inline prov/gni/src/gnix_buddy_allocator.c:254]: Why do you add newlines to some of the print statements (and then why not to others)?

File: prov/gni/src/gnix_buddy_allocator.c [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+	if (!alloc_handle) {
+		error = strerror_r(errno, err_buf, sizeof(err_buf));
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Could not create buddy allocator handle.",
+			  error);
+		return -FI_ENOMEM;
+	}
+
+	fastlock_init(&alloc_handle[0]->lock);
+	fastlock_acquire(&alloc_handle[0]->lock);
```
Comment (ztiffany) [inline prov/gni/src/gnix_buddy_allocator.c]: You shouldn't need to lock the not yet created allocator.

File: prov/gni/include/gnix_buddy_allocator.h [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+#include "fi_list.h"
+#include "gnix_bitmap.h"
+#include "gnix_util.h"
+#include <math.h>
+#include <stdlib.h>
+
+#define MIN_BLOCK_SIZE 16
+
+/* evaluates to zero if X is not a power of two, otherwise evaluates to X - 1 */
+#define IS_NOT_POW_TWO(X) (((X) & (~(X) + 1)) ^ (X))
```
Comment (jswaro) [inline prov/gni/include/gnix_buddy_allocator.h:70]: You might consider the following as well. It should result in fewer operations. 

IS_POW_TWO(v) v && (v & (v - 1)))
IS_NOT_POW_TWO(v) !IS_POW_TWO(v)

File: prov/gni/include/gnix_buddy_allocator.h [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+ */
+#define BUDDY(X, LEN, BASE) (void *) ((((size_t) (BASE) - (size_t) (X)) /\
+				       (LEN)) % 2 ? (size_t) (X) - (LEN) :\
+				      (size_t) (X) + (LEN))
+
+/* Calculate the offset of a free block, OFFSET = MIN * 2^MULT. */
+#define OFFSET(MIN, MULT) ((MIN) * (1 << (MULT)))
+
+/* Find the index into the free list with block size LEN. */
+#define LIST_INDEX(LEN, MIN) (size_t) (log2((LEN) / (double) (MIN)))
```
Comment (jswaro) [inline prov/gni/include/gnix_buddy_allocator.h]: I noticed that you use log2 here. glibc's log functions have gotten better over the years, but they have been notoriously slow in the past. What assumptions can you make with len and min? You may be able to find a better solution and remove the dependency on libm. 

See http://graphics.stanford.edu/~seander/bithacks.html#IntegerLogDeBruijn for possible solutions if your assumptions fit.

File: prov/gni/include/gnix_buddy_allocator.h [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+/* Find the address of X's buddy block:
+ * If the "index" of block X is even then the buddy must be to the right of X,
+ * otherwise the buddy is to the left of X.
+ */
+#define BUDDY(X, LEN, BASE) (void *) ((((size_t) (BASE) - (size_t) (X)) /\
+				       (LEN)) % 2 ? (size_t) (X) - (LEN) :\
+				      (size_t) (X) + (LEN))
+
+/* Calculate the offset of a free block, OFFSET = MIN * 2^MULT. */
+#define OFFSET(MIN, MULT) ((MIN) * (1 << (MULT)))
```
Comment (jswaro) [inline prov/gni/include/gnix_buddy_allocator.h]: I'm surprised this is ok. I'm not sure how this would react if this header was pulled into a file that defined a macro named 'MIN'.

File: prov/gni/src/gnix_buddy_allocator.c [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+ * track of in a doubly linked list, or free list.  Here are the structures
+ * and format of the data structures used in the buddy allocator.  For
+ * a description of each field please see gnix_buddy_allocator.h.
+ *
+ 		Handle structure:
+		┌──────┬──────┬─────┬─────┬────────┬───────┐
+		│ BASE │ len  │ min │ max │ nlists │ LISTS │
+		└──────┴──────┴─────┴─────┴────────┴───────┘
+ * The LISTS pointer points to an array of dlist structures, each containing a
+ * head pointer to the begging of a free list.  Note that the first element of
```
Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c]: beginning

File: prov/gni/src/gnix_buddy_allocator.c [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+
+	dlist_remove(tmp = alloc_handle->lists[j].next);
+
+	_gnix_set_bit(&alloc_handle->bitmap,
+		      BITMAP_INDEX(tmp, alloc_handle->base, MIN_BLOCK_SIZE,
+				   OFFSET(MIN_BLOCK_SIZE, j)));
+
+
+	/* Split the block until we reach list "i" */
+	for (j--; j > i; j--) {
```
Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c:165]: Could you speculate what the most common difference between j and i will be? 1, 2, 4? If the value is high enough, you might find a linear set of the bitset to be prohibitively slow.

File: prov/gni/src/gnix_buddy_allocator.c [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+				 handle_t **alloc_handle)
+{
+	char err_buf[256] = {0}, *error = NULL;
+	int fi_errno;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	/* Ensure parameters are valid */
+	if (!base || !len || !max || max > len || !alloc_handle ||
+	    IS_NOT_POW_TWO(max) || (len % max) || !(len / MIN_BLOCK_SIZE * 2)) {
```
Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c]: There are a large amount of operations here that are very infrequently necessary. This is probably a good candidate for a unlikely macro. It seems like the buddy allocator is something we'll only use internally, so errors should be very infrequent (we are prone to mistakes however). You can warn the branch predictor that this will likely not be true, so it might buffer some of the other operations while processing these.

File: prov/gni/src/gnix_buddy_allocator.c (hunk 3 above)
Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c]: agreed

File: prov/gni/src/gnix_buddy_allocator.c [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+	alloc_handle[0]->max = max;
+
+	if (__gnix_buddy_create_lists(alloc_handle[0])) {
+		fastlock_release(&alloc_handle[0]->lock);
+		free(*alloc_handle);
+		return -FI_ENOMEM;
+	}
+
+	if ((fi_errno = _gnix_alloc_bitmap(&alloc_handle[0]->bitmap,
+					   len / MIN_BLOCK_SIZE * 2))) {
```
Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c:283]: why len / MBS *2? Maybe a comment here to explain your calculation

File: prov/gni/src/gnix_buddy_allocator.c [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+			  "Could not allocate buddy block.\n");
+		return -FI_ENOMEM;
+	}
+
+	/*
+	 * Remove the block from opposite sides of adjacent lists to reduce
+	 * fragmentation and improve coalescing
+	 */
+	if (i % 2) {
+		dlist_remove(*ptr = alloc_handle->lists[i].prev);
```
Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c]: Nit: 
I'd rather not have expressions passed to functions. This complicates things when you have a function that is later turned into a macro as the expression may be executed more than once. 

Example:
POW2(x) ((x) \* (x))
z = 2
Y = POW(z++);

Y should have been 4, but will be 6

I realize the statement I'm referring to isn't technically an expression, but it still remains an assignment, and I'd rather those not be within a function call either.

Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c]: While we are on the topic of funky looking things, you could do this to avoid the conditional ... but don't 

> _ptr = (struct dlist_entry *) ((struct dlist_entry *_) &alloc_handle->lists[i]) + (i & 1)); 

readability is far more important than that. Typically, avoiding a conditional will save the CPU's branch predictor from making any unnecessary assumptions, but it doesn't always pan out in your favor.

File: prov/gni/include/gnix_buddy_allocator.h (hunk 1 above)
Comment (e-harvey) [inline prov/gni/include/gnix_buddy_allocator.h]: Ok, I agree.

File: prov/gni/src/gnix_buddy_allocator.c (hunk 2 above)
Comment (e-harvey) [inline prov/gni/src/gnix_buddy_allocator.c:254]: My mistake.

File: prov/gni/src/gnix_buddy_allocator.c (hunk 3 above)
Comment (e-harvey) [inline prov/gni/src/gnix_buddy_allocator.c]: Okay.

File: prov/gni/include/gnix_buddy_allocator.h (hunk 4 above)
Comment (e-harvey) [inline prov/gni/include/gnix_buddy_allocator.h:70]: Could you elaborate?

File: prov/gni/include/gnix_buddy_allocator.h (hunk 5 above)
Comment (e-harvey) [inline prov/gni/include/gnix_buddy_allocator.h]: Ok, great.  Both Len and min are powers of two, and min <= Len.

I'll have to figure out the 64bit solution to this.

File: prov/gni/include/gnix_buddy_allocator.h (hunk 6 above)
Comment (e-harvey) [inline prov/gni/include/gnix_buddy_allocator.h]: I will rename MIN.

File: prov/gni/src/gnix_buddy_allocator.c (hunk 7 above)
Comment (e-harvey) [inline prov/gni/src/gnix_buddy_allocator.c]: Ok

File: prov/gni/src/gnix_buddy_allocator.c (hunk 8 above)
Comment (e-harvey) [inline prov/gni/src/gnix_buddy_allocator.c:165]: I'm not sure I understand.  Perhaps I need to rethink the bitmap layout.  From what I understand the buddy allocator should be managing a base block of at most 32MB.

Assuming the largest block size is 16MB and the smallest block size is 16 bytes, the largest difference between j and i would be 20 since:
i = 0 (the index of the 16byte block list)
j = 20 (the index of the 16MB block list).

File: prov/gni/src/gnix_buddy_allocator.c (hunk 9 above)
Comment (e-harvey) [inline prov/gni/src/gnix_buddy_allocator.c]: Okay, I agree.

File: prov/gni/src/gnix_buddy_allocator.c (hunk 10 above)
Comment (e-harvey) [inline prov/gni/src/gnix_buddy_allocator.c:283]: This should be 2 \* (Len / MBS).

Len is the number of bytes that the base block spans.

To have enough bits to flag each block as allocated or split whether the block size is 16B, 32B, 64B, ... we need:

Len / 16 bits for all the 16B blocks
Len / 32 bits for all the 32B blocks
Len / 64 bits for all the 64B blocks
...
Len / MaxBlockSize bits

So, as a maximum (if the max block size = Len) we need one less than two times Len / 16.

File: prov/gni/src/gnix_buddy_allocator.c (hunk 11 above)
Comment (e-harvey) [inline prov/gni/src/gnix_buddy_allocator.c]: Okay, that makes sense.

Comment (e-harvey) [inline prov/gni/src/gnix_buddy_allocator.c]: That works.  I agree.

File: prov/gni/include/gnix_buddy_allocator.h (hunk 4 above)
Comment (jswaro) [inline prov/gni/include/gnix_buddy_allocator.h:70]: My mistake. The functions are basically equivalent. The same amount of operations, but assuming that a bitwise XOR and a ! are of the same speed, there is no difference.

File: prov/gni/src/gnix_buddy_allocator.c (hunk 8 above)
Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c:165]: I see. I misread what you were doing here. It seems fine to me.

Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c:165]: Perhaps you could explain your use of the bitmaps a bit better to me ?

Comment (e-harvey) [inline prov/gni/src/gnix_buddy_allocator.c:165]: Sure, the bitmap in the handle_t struct has a bit for every block size, the smallest blocks are mapped to in the first half of the bitmap and the rest of the blocks are mapped to in the second half of the bitmap.  The bitmap is over allocated and will always waste at least one bit.

A bit (corresponding to a block of a certain size) is set when that block is allocated or split.

For example, if the length of the base block is 128B,
MinBlockSize = 16B,
MaxBlockSize = 64B then bitmap looks like:

16|16|16|16|16|16|16|16|32|32|32|32|64|64

So, all the 16s are bits mapping to all the blocks of size 16 that we can fit in our base block; this goes for the 32s and 64s as well.  We are not using the 2 hidden bits after the last 64.

When freeing, we try to coalesce the free block with a buddy block; the two blocks must be the same size.  We can find the state of any buddy block quickly with BITMAP_INDEX and if that bit is set, then we have to insert the free block into its free list without coalescing.  If the bit is not set, we coalesce until we reach a buddy block that is allocated or split, or we reach the MaxBlockSize list, then we insert the block into its free list.

So, by using this bitmap layout we avoid searching through our free lists but waste some space.

File: prov/gni/include/gnix_buddy_allocator.h [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+			      (LEN))
+
+/* Find the bitmap index for block X */
+#define BITMAP_INDEX(X, BASE, MIN, LEN) (size_t) ((size_t) ((X) - (BASE)) /\
+					 (MIN) + 2 * log2((LEN) / (MIN)))
+
+/*
+ * The following macro doesn't work when the base address starts at zero.
+ * #define BUDDY(X, SIZE_X, BASE) (void *) ((size_t) (X) ^ (SIZE_X))
+ */
```
Comment (e-harvey) [inline prov/gni/include/gnix_buddy_allocator.h]: Correction:  this macro doesn't work when the base block starts at an address with the SIZE_X bit set.  When this happens the buddy block of the first block in the base block is calculated to be less than the starting address of the base address which is out of the address range that the base block covers.

For example,
alloc_handle->base = BASE = 0x7fffde1d
SIZE_X = 0x10
X = 0x7fffde1d

BUDDY(0x7fffde1d, 0x10, 0x7fffde1d) =

0x10 XOR 0x7fffde1d =

0x7fffde0d

File: prov/gni/src/gnix_buddy_allocator.c (hunk 8 above)
Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c:165]: I suppose I'm not worried about the space usage anymore, but I'm still not sure how it is that you use the bitmap and how you get from 0's and 1's to 16's and 32's .

File: prov/gni/include/gnix_buddy_allocator.h (hunk 5 above)
Comment (jswaro) [inline prov/gni/include/gnix_buddy_allocator.h]: Your len will never be over 4GB, so I don't see why the uint32_t version would not work.

File: prov/gni/src/gnix_buddy_allocator.c (hunk 8 above)
Comment (e-harvey) [inline prov/gni/src/gnix_buddy_allocator.c:165]: Okay, the bitmap is setup like this:

All the blocks that the buddy allocator allocates from the base block are of size X, where X = MBS \* 2^Z, MBS is the minimum block size and Z is a non-negative integer.

The bitmap has 2 \* (Len / MBS) bits and it's setup so that the first Len / MBS bits in the bitmap flag each block of size MBS as free or allocated.

Len is the number of bytes in the base block.
The base block is pointed to by void *base.

The first bit in the bitmap flags the first block of size MBS.  The first block of size MBS uses the address range:
base to (base + MBS - 1).

The second bit in the bitmap flags the second block of size MBS.  The second block of size MBS uses the address range:
(base + MBS) to (base + 2 \* MBS - 1)

The third bit in the bitmap flags the third block of size MBS.  The third block of size MBS uses the address range:
(base + 2 \* MBS) to (base + 3 \* MBS - 1)

And so on until we reach the Len / MBS bit in the bitmap.

The second Len / MBS bits in the bitmap flag the remaining blocks of size X as free allocated, or split where X > MBS.

So, the first bit in the second Len / MBS bits in the bitmap flags the first block of size MBS \* 2.  The first block of size MBS \* 2 uses the address range:
base to (base + MBS \* 2 - 1)

And so on until we reach the next block size.

The bitmap is used like this:

A bit is set to 1 when a block is allocated, or when the block is split into two smaller blocks.

A bit is reset to 0 when a block is freed, or when a free block is coalesced with another already free and unsplit block.

The bitmap is only read for coalescing blocks. When a block Y is freed we look at the bit in the bitmap for the buddy block of Y, if that bit is set then the buddy of Y is allocated, split, or both in which case we cannot coalesce Y with its buddy block.  However, if the bitmap bit for the buddy of Y is reset, then the buddy block of Y is free and not split, so we coalesce Y with the buddy of block of Y and continue to coalesce this new larger block with its buddy block until we reach the max block size or a buddy block that is allocated, split, or both.

I hope that makes sense, let me know if it doesn't.

File: prov/gni/include/gnix_buddy_allocator.h (hunk 5 above)
Comment (e-harvey) [inline prov/gni/include/gnix_buddy_allocator.h]: That makes sense.  Great.

File: prov/gni/src/gnix_buddy_allocator.c (hunk 8 above)
Comment (jswaro) [inline prov/gni/src/gnix_buddy_allocator.c:165]: > A bit is set to 1 when a block is allocated, or when the block is split into two smaller blocks.
> If a block is split, how would you know that it is allocated or not? It seems as though the bit is being overloaded. 
> Perhaps this is something we could discuss over a phone call for the sake of expediency.

File: prov/gni/include/gnix_buddy_allocator.h [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+ * Free a previously allocated buffer
+ *
+ * @param[in] alloc_handle 	Previously allocated GNI buddy_alloc_handle to
+ * use as allocator.
+ *
+ * @param[in/out] ptr		Pointer to an address where the address of the
+ * allocated buffer will be returned.
+ *
+ * @param[in] len		Size of buffer to allocate in bytes.
+ *
```
Comment (hppritcha) [inline prov/gni/include/gnix_buddy_allocator.h:193]: I'd suggest rewording the description of len to something like:
"size of previously allocated buffer in byte"

File: prov/gni/include/gnix_buddy_allocator.h [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+ * @return FI_SUCCESS		Upon successfully allocating a buffer.
+ *
```
Comment (hppritcha) [inline prov/gni/include/gnix_buddy_allocator.h:195]: Upon successful freeing of buffer.
